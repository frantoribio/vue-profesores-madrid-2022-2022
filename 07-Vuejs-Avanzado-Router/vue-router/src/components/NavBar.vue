<template>
  <nav>
    <RouterLink to="/">Home</RouterLink>
    <RouterLink to="/about">About</RouterLink>
    <!-- Si quiero usar el nombre de las vistas y no su url-->
    <RouterLink :to="{ name: 'home' }">Home Named</RouterLink>
    <!-- Externo a Google -->
    <RouterLink :to="{ name: 'google' }">Google</RouterLink>
    <!-- Autenticada-->
    <RouterLink :to="{ name: 'autenticada' }">Autenticada</RouterLink>
  </nav>
</template>

<script setup>
  import { RouterLink } from 'vue-router'
</script>

<style scoped>
  nav {
    padding: 0 30px;
  }
  nav a {
    color: #2c3e50;
    font-weight: bold;
    margin: 0 10px;
  }
  nav a {
    cursor: pointer;
  }
  /* Para el enlace activo */
  .router-link-exact-active {
    color: #42b983;
  }
</style>
