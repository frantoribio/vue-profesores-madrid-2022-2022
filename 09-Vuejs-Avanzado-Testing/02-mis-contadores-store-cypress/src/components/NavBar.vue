<template>
  <nav>
    <RouterLink
      :to="{ name: 'Home' }"
      data-cy="navbar-home"
      >Home</RouterLink
    >
    <RouterLink
      :to="{ name: 'About' }"
      data-cy="navbar-about"
      >About</RouterLink
    >
  </nav>
</template>

<script setup>
  import { RouterLink } from 'vue-router'
</script>

<style scoped>
  nav {
    width: 100%;
    font-size: 12px;
    text-align: center;
    margin-top: 2rem;
  }

  nav a.router-link-exact-active {
    color: var(--color-text);
  }

  nav a.router-link-exact-active:hover {
    background-color: transparent;
  }

  nav a {
    display: inline-block;
    padding: 0 1rem;
    border-left: 1px solid var(--color-border);
  }

  nav a:first-of-type {
    border: 0;
  }

  nav {
    text-align: left;
    margin-left: -1rem;
    font-size: 1rem;

    padding: 1rem 0;
    margin-top: 1rem;
  }
</style>
