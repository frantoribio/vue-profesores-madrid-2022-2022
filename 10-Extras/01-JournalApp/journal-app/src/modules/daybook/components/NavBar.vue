<template>
  <div class="navbar rounded-md bg-primary text-primary-content shadow-md">
    <div class="navbar-start">
      <button
        data-testid="navbar-button-home"
        class="btn btn-ghost text-xl normal-case"
        @click="goToHome()"
      >
        <img
          src="@/assets/images/logo.png"
          alt="logo"
          class="mr-3 inline-block w-10 text-right"
        />
        DayBook
      </button>
    </div>
    <div class="navbar-end">
      <button
        data-testid="navbar-button-logout"
        v-if="userStore.isLoggedIn"
        class="btn btn-primary gap-2"
        @click="logOut()"
      >
        {{ userStore.user.username }}
        <Icon
          icon="material-symbols:exit-to-app-rounded"
          :inline="true"
          class="h-8 w-8"
        />
      </button>
      <button
        data-testid="navbar-button-login"
        v-else
        class="btn btn-primary gap-2"
        @click="logIn()"
      >
        <Icon
          icon="bxs:user"
          :inline="true"
          class="h-8 w-8"
        />
      </button>
    </div>
  </div>
</template>

<script setup>
  import UserStore from '@/stores/users'
  import { Icon } from '@iconify/vue'
  import { useRouter } from 'vue-router'

  const userStore = new UserStore()

  const logOut = async () => {
    await userStore.logOut()
  }

  const logIn = async () => {
    await userStore.logIn()
  }

  const router = useRouter()
  // const goToDaybook = () => router.push({ name: 'daybook-no-entry' })
  const goToHome = () => router.push({ name: 'home' })
</script>

<style lang="scss" scoped></style>
