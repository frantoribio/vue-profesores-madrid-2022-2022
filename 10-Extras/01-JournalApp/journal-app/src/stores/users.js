import users from '@/data/users'
import espera from '@/helpers/espera'
import JournalStore from '@/modules/daybook/stores/journal'
import { defineStore } from 'pinia'

export const UserStore = defineStore({
  id: 'UserStore',
  state: () => ({
    user: null,
  }),

  getters: {
    isLoggedIn: (state) => state.user !== null,
  },

  actions: {
    setUser(user) {
      this.user = user
    },

    async logIn() {
      await espera(500)
      // TODO: Compruebas la sesi√≥n en remoto o Firebase
      this.setUser(users[0])
      // Le cargamos las entradas del usuario!!!
      const journalStore = JournalStore()
      journalStore.fetchEntries()
    },

    async logOut() {
      await espera(1000) // Solo por perder el tiempo y parezca asincrono
      // Le limpiamos los datos del usuario!!!
      const journalStore = JournalStore()
      journalStore.clearData()
      this.setUser(null)
    },
  },
})

export default UserStore
