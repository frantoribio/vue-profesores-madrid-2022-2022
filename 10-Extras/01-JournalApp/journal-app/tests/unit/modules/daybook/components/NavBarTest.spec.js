import { beforeEach, describe, expect, test } from 'vitest'

import { createTestingPinia } from '@pinia/testing'
import { shallowMount } from '@vue/test-utils'
import { createRouterMock, injectRouterMock } from 'vue-router-mock'

import NavBar from '@/modules/daybook/components/NavBar.vue'

import users from '../stores/mocks/users.mocks'

// Para esperar las promesas pendientes tambien puedes usar: https://www.npmjs.com/package/flush-promises
// import { setImmediate } from 'timers'
// const flushPromises = () => new Promise(setImmediate)

describe('Daybook Componente -> NavBar', () => {
  let wrapper
  const router = createRouterMock()

  beforeEach(() => {
    injectRouterMock(router)
    wrapper = shallowMount(NavBar, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              UserStore: { user: users[0] },
            },
          }),
        ],
      },
    })
  })

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.text()).toMatchSnapshot()
  })

  test('Debe mostrar el nombre del usuario', () => {
    const button = wrapper.find('[data-testid="navbar-button-logout"]')
    expect(button.text()).toBe(users[0].username)
  })

  test('Debe llevar a home al pulsar el boton', async () => {
    const button = wrapper.find('[data-testid="navbar-button-home"]')
    expect(button.text()).toBe('DayBook')
    await button.trigger('click')
    expect(router.push).toHaveBeenCalledWith({ name: 'home' })
  })

  test('Debe hacer logout', async () => {
    const button = wrapper.find('[data-testid="navbar-button-logout"]')
    expect(button.text()).toBe(users[0].username)
    await button.trigger('click')
    expect(wrapper.vm.userStore.logOut).toHaveBeenCalled()
    // await flushPromises()
  })

  test('Debe hacer login', async () => {
    wrapper = shallowMount(NavBar, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              UserStore: { user: null },
            },
          }),
        ],
      },
    })
    const button = wrapper.find('[data-testid="navbar-button-login"]')
    await button.trigger('click')
    expect(wrapper.vm.userStore.logIn).toHaveBeenCalled()
    // await flushPromises()
  })
})
