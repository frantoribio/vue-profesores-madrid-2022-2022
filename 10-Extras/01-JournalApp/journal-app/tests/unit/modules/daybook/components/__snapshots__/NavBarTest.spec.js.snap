// Vitest Snapshot v1

exports[`Daybook Componente -> NavBar > Debe hacer match con el snapshot 1`] = `"DayBook jdoe"`;
